File: packages/todo/lib/src/model/todo.dart
import 'package:core/core.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'todo.freezed.dart';

class TodoId extends ValueId<int> {
  const TodoId(int value) : super(value);
}

@freezed
class Todo with _$Todo {
  const factory Todo({
    required TodoId? id,
    required String title,
    required String description,
    required Set<String> tags,
    required bool isCompleted,
    required DateTime createdAt,
    required DateTime updatedAt,
  }) = _Todo;

  /// Create a new Todo with default values.
  factory Todo.empty() {
    return Todo(
      id: null,
      title: '',
      description: '',
      tags: {},
      isCompleted: false,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
    );
  }
}

File: todo-api/app/io/todo/metadata/todo/domain/models/Todo.scala
package io.todo.metadata.todo.domain.models

import org.joda.time.DateTime

case class Todo(
    id: TodoId,
    title: String,
    description: String,
    tags: Set[String],
    completed: Boolean,
    parentId: Option[TodoId],
    deletedAt: Option[org.joda.time.DateTime],
    createdAt: org.joda.time.DateTime,
    updatedAt: org.joda.time.DateTime
)

object Todo {
  def patch(newObject: Todo, oldObject: Todo): Todo =
    newObject.copy(
      id = oldObject.id,
      createdAt = oldObject.createdAt,
      updatedAt = DateTime.now
    )
}
